primitives/fee-handler/src/fee_handler.rs

```diff
@@ -0,0 +1,367 @@
+use crate::withdraw_credit::WithdrawCredit;
+use core::marker::PhantomData;
+use frame_support::{
+	pallet_prelude::DispatchError,
+	traits::{fungible, fungibles, ConstU32, Defensive, Imbalance},
+	BoundedVec,
+	__private::log,
+};
+use parity_scale_codec::{Decode, Encode};
+use std::fmt::Debug;
+
+/// Distributes shares of a base fee to some beneficiaries.
+pub trait DistributeFee {
+	/// AccountId type used.
+	type AccountId;
+
+	/// Scalar balance type.
+	type Balance;
+
+	/// Fee identifier used to derive the fee distribution.
+	type FeeIdentifier;
+
+	/// Maximum number of distributions of shares from the base fee to beneficiaries.
+	type MaxDistributions;
+
+	fn distribute_fee(
+		base_fee: Self::Balance,
+		account: &Self::AccountId,
+		identifier: &Self::FeeIdentifier,
+	) -> Option<BoundedVec<Payment<Self::AccountId, Self::Balance>, Self::MaxDistributions>>;
+}
+
+/// Payment to be executed.
+#[derive(Debug, Encode, Decode, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct Payment<AccountId, Balance> {
+	beneficiary: AccountId,
+	amount: Balance,
+}
+
+impl<AccountId, Balance> Payment<AccountId, Balance> {
+	pub fn new(beneficiary: AccountId, amount: Balance) -> Self {
+		Self { beneficiary, amount }
+	}
+}
+
+/// Abstraction of withdrawing fees in one asset and return allocating the fees in the same or
+/// another asset.
+pub trait FeeHandler {
+	type AccountId;
+
+	type AssetId;
+
+	/// Scalar type of the fee balance.
+	type Balance;
+
+	type AffiliateFeeIdentifier;
+	type TournamentFeeIdentifier;
+
+	/// Withdraws the `base_fee` denominated in `payment_asset` allocate shares of the base fee to
+	/// the affiliate, tournament, and treasury if implemented.
+	fn withdraw_and_pay_fees(
+		payer: &Self::AccountId,
+		payment_asset: Self::AssetId,
+		base_fee: Self::Balance,
+		tournament_id: &Self::TournamentFeeIdentifier,
+		affiliate_id: &Self::AffiliateFeeIdentifier,
+		treasury_pot: &Self::AccountId,
+	) -> Result<(), DispatchError>;
+
+	/// Withdraws the `base_fee` denominated in `payment_asset` allocate it fully to the
+	/// `treasury_pot`.
+	fn withdraw_and_deposit_into_treasury(
+		who: &Self::AccountId,
+		asset_id: Self::AssetId,
+		treasury_pot: &Self::AccountId,
+		amount: Self::Balance,
+	) -> Result<(), DispatchError>;
```

Comment:
This is subject to change. For the current intermediate state, I thought it is the easiest if we just supply the account id. Only later I noticed that AAA does some accounting in its storage, unfortunately. As mentioned before, I will look at it this week.